compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -1172,6 +1149,341 @@ LogicalResult createAPIFunctions(IREE::VM::ModuleOp moduleOp,
+    functions +=
+        "static const iree_vm_native_function_ptr_t " + functionName + "[] = {";
+    if (exportedFunctions.empty()) {
+      // Empty list placeholder.
+      functions += "{0},";
+    } else {
+      // We only add exported functions to the table, as calls to internal
+      // functions are directly mapped to C function calls of the generated
+      // implementation.
+      for (auto funcOp : exportedFunctions) {
+        auto funcName = funcOp.getName();
+        functions += std::string("{") +
+                     "(iree_vm_native_function_shim_t)iree_emitc_shim, " +
+                     "(iree_vm_native_function_target_t)" + funcName.str() +
+                     "},";
+      }
+    }
+    functions += "};";
+    builder.create<emitc::VerbatimOp>(loc, functions);
+
+    // module descriptor
+    // TODO(simon-camp): support module-level reflection attributes
+    std::string descriptorName = moduleOp.getName().str() + "_descriptor_";
+    std::string descriptor;
+    descriptor +=
+        "static const iree_vm_native_module_descriptor_t " + descriptorName +
+        " = {"
+        // name:
+        + printStringView(moduleOp.getName()) +
+        ","
+        // version:
+        + std::to_string(moduleOp.getVersion().value_or(0u)) +
+        ","
+        // attrs:
+        + "0," +
+        "NULL,"
+        // dependencies:
+        + std::to_string(dependencies.size()) + "," + dependenciesName +
+        ","
+        // imports:
+        + std::to_string(importOps.size()) + "," + importName +
+        ","
+        // exports:
+        + std::to_string(exportedFunctions.size()) + "," + exportName +
+        ","
+        // functions:
+        + std::to_string(exportedFunctions.size()) + "," + functionName + "," +
+        "};";
+
+    builder.create<emitc::VerbatimOp>(loc, descriptor);
+
+    // move functions marked as emitAtEnd to the end of the module
+    auto funcs =
+        SmallVector<mlir::func::FuncOp>(moduleOp.getOps<mlir::func::FuncOp>());
+    for (auto func : funcs) {
+      if (typeConverter.analysis.lookupFunction(func).shouldEmitAtEnd()) {
+        func->moveBefore(moduleOp.getBlock().getTerminator());
+      }
+    }
+
+    builder.setInsertionPoint(moduleOp.getBlock().getTerminator());
+    emitc_builders::preprocessorDirective(builder, loc, emitc_builders::ENDIF,
+                                          "  // EMITC_IMPLEMENTATION");
+
+    // insert a verbatim op before private functions
+    for (auto func : moduleOp.getOps<mlir::func::FuncOp>()) {
+      emitc_builders::makeFuncStatic(builder, loc, func);
+    }
+  }
+
+  for (auto op : opsToRemove) {
+    op->erase();
+  }
+
+  // TODO(simon-camp): The emitc code emitter expects a builtin.module as the
+  // outer container of the supported operations. Instead of nesting a
+  // builtin.module inside the vm.module here we should change this conversion
+  // pass to run on the outer builtin.module and replace the vm.module
+  // completely.
```

Comment:
```suggestion
   // TODO(simon-camp): The Cpp Emitter expects a builtin.module as the
   // outer container of the supported operations. Instead of nesting a
   // builtin.module inside the vm.module as in the current implementation,
   // the conversion pass should be changed to replace the vm.module
   // with a builtin.module.
```